packages/tenant-process/src/services/readModelService.ts

```diff
@@ -18,13 +22,153 @@ import {
+    )
+    .exhaustive();
+
+const makeAttributesFilter = (
+  fieldName: Extract<
+    AgreementDataFields,
+    "certifiedAttributes" | "declaredAttributes" | "verifiedAttributes"
+  >,
+  attributeIds: AttributeId | AttributeId[]
+): ReadModelFilter<Agreement> | undefined =>
+  match(attributeIds)
+    .with(P.string, (id) => ({
+      [`data.${fieldName}`]: { $elemMatch: { id } },
+    }))
+    .with(P.array(P.string), (ids) =>
+      ids.length === 0
+        ? undefined
+        : {
+            [`data.${fieldName}`]: {
+              $elemMatch: { id: { $in: ids } },
+            },
+          }
+    )
+    .exhaustive();
+
+const getAgreementsFilters = (
+  filters: AgreementQueryFilters
+): { $match: object } => {
+  const upgradeableStates = [
+    agreementState.draft,
+    agreementState.active,
+    agreementState.suspended,
+  ];
+
+  const {
+    attributeId,
+    producerId,
+    consumerId,
+    eserviceId,
+    descriptorId,
+    agreementStates,
+    showOnlyUpgradeable,
+  } = filters;
+
+  const agreementStatesFilters = match(agreementStates)
+    .with(P.nullish, () => (showOnlyUpgradeable ? upgradeableStates : []))
+    .with(
+      P.when(
+        (agreementStates) => agreementStates.length === 0 && showOnlyUpgradeable
+      ),
+      () => upgradeableStates
+    )
+    .with(
+      P.when(
+        (agreementStates) => agreementStates.length > 0 && showOnlyUpgradeable
+      ),
+      (agreementStates) =>
+        upgradeableStates.filter((s) => agreementStates.includes(s))
+    )
+    .otherwise((agreementStates) => agreementStates);
+
+  const queryFilters = {
+    ...makeFilter("producerId", producerId),
+    ...makeFilter("consumerId", consumerId),
+    ...makeFilter("eserviceId", eserviceId),
+    ...makeFilter("descriptorId", descriptorId),
+    ...(agreementStatesFilters &&
+      agreementStatesFilters.length > 0 && {
+        "data.state": {
+          $in: agreementStatesFilters.map((s) => s.toString()),
+        },
+      }),
+    ...(attributeId && {
+      $or: [
+        makeAttributesFilter("certifiedAttributes", attributeId),
+        makeAttributesFilter("verifiedAttributes", attributeId),
+        makeAttributesFilter("declaredAttributes", attributeId),
+      ],
+    }),
+  };
+  return { $match: queryFilters };
+};
+
+const getAllAgreements = async (
+  agreements: AgreementCollection,
+  filters: AgreementQueryFilters
+): Promise<Agreement[]> => {
+  const data = await agreements
+    .aggregate([getAgreementsFilters(filters)])
+    .toArray();
+
+  const result = z
+    .array(
+      z.object({
+        data: Agreement,
+      })
+    )
+    .safeParse(data);
+
+  if (!result.success) {
+    logger.error(
+      `Unable to parse agreements items: result ${JSON.stringify(
+        result
+      )} - data ${JSON.stringify(data)} `
+    );
+    throw genericError("Unable to parse agreements items");
+  }
+  return result.data.map((d) => d.data);
+};
+
```

Comment:
Please double-check if the agreements query can be simplified. Maybe we don't need such flexibility for this use case.